protocols/luci-proto-bonding/htdocs/luci-static/resources/protocol/bonding.js

```diff
@@ -0,0 +1,442 @@
+	}
+}
+
+function validate_arp_policy(section, value) {
+
+	var opt = this.map.lookupOption('link_monitoring', section);
+	var selected_link_monitoring = opt[0].formvalue(section);
+
+	var opt = this.map.lookupOption('bonding_policy', section);
+	var selected_policy = opt[0].formvalue(section);
+
+	if (selected_link_monitoring == 'arp') {
+		if (selected_policy == '802.3ad' || selected_policy == 'balance-tlb' || selected_policy == 'balance-alb') {
+			return _('ARP monitoring is not supported for the selected policy!');
+		}
+	}
+
+	return true;
+}
+
+function validate_arp_ip_targets(section, value) {
+
+	var opt = this.map.lookupOption('link_monitoring', section);
+	var selected_link_monitoring = opt[0].formvalue(section);
+
+	var opt = this.map.lookupOption('arp_ip_target', section);
+	var selected_arp_ip_targets = opt[0].formvalue(section);
+
+	var opt = this.map.lookupOption('bonding_policy', section);
+	var selected_policy = opt[0].formvalue(section);
+
+	if (selected_link_monitoring == 'arp' && selected_arp_ip_targets.length == 0) {
+		return _('You must select at least one ARP IP target if ARP monitoring is selected!');
+	}
+
+	return true;
+}
+
+function validate_primary_interface(section, value) {
+
+	var primary_valid = 'false';
+
+	var opt = this.map.lookupOption('bonding_policy', section);
+	var selected_policy = opt[0].formvalue(section);
+
+	var opt = this.map.lookupOption('slaves', section);
+	var selected_slaves = opt[0].formvalue(section);
+
+	var opt = this.map.lookupOption('primary', section);
+	var selected_primary = opt[0].formvalue(section);
+
+	if (selected_policy == 'active-backup' || selected_policy == 'balance-tlb' || selected_policy == 'balance-alb') {
+
+		for (var i = 0; i < selected_slaves.length; i++) {
+			if (selected_slaves[i] == selected_primary) {
+				primary_valid = 'true';
+			}
+		}
+
+		if (primary_valid == 'false') {
+			return _('You must select a primary interface which is included in selected slave interfaces!');
+		}
+	}
+
+	return true;
+}
+
+return network.registerProtocol('bonding', {
+	getI18n: function() {
+		return _('Link Aggregation (Channel Bonding)');
+	},
+
+	getIfname: function() {
+		return null;
+	},
+
+	getOpkgPackage: function() {
+		return 'bonding';
+	},
+
+	isFloating: function() {
+		return true;
+	},
+
+	isCreateable: function(ifname) {
+		return getSelectableSlaves(ifname).then(L.bind(function(devices) {
+			return devices.length == 0 ?  _('No more slaves available') : null;
+		}, this));
+
+		return _('No more slaves available');
+	},
+
+	isVirtual: function() {
+		return true;
+	},
+
+	getDevices: function() {
+		return null;
+	},
+
+	containsDevice: function(ifname) {
+		return (network.getIfnameOf(ifname) == this.getIfname());
+	},
+
+	renderFormOptions: function(s) {
+		var o;
+
+		o = s.taboption('general', form.Value, 'ipaddr',
+				_('IPv4 address'),
+				_('The local IPv4 address'));
+		o.datatype = 'ip4addr';
+		o.validate = validateEmpty;
```

Comment:
You can replace this and all other `o.validate = validateEmpty` lines with `o.rmempty = false` to achieve the same effect which allows you to get rid of `function validateEmpty() { ... }`

Reply:
Oh, this is really a valuable tip I didn't know.
It works well, but it misbehaves a little if you add some o.value() without an o.default().
In this case yo get a dropdown with "-- Please choose --" per default but this is not interpreted as empty (leading e.g. to create a new interface nevertheless)

Reply:
Well, I removed all o.validate and substituted them with o.rmempty exept the special case I mentioned above.